File: internal/core/agentloop_test.go
package core

import (
	"context"
	"strings"
	"testing"

	"github.com/overkazaf/re-agent/internal/types"
)

// scriptedProvider replays a fixed list of responses, one per turn.
type scriptedProvider struct {
	config    *types.ProviderConfig
	responses []types.ProviderResponse
	turn      int
	seen      [][]types.Message
	systems   []string
	tools     [][]types.Tool
	fresh     []bool
}

func (p *scriptedProvider) Name() string                  { return "scripted" }
func (p *scriptedProvider) Config() *types.ProviderConfig { return p.config }

func (p *scriptedProvider) Complete(input types.ProviderInput) (types.ProviderResponse, error) {
	p.seen = append(p.seen, input.Messages)
	p.systems = append(p.systems, input.System)
	p.tools = append(p.tools, input.Tools)
	p.fresh = append(p.fresh, input.FreshSession)
	index := p.turn
	if index > len(p.responses)-1 {
		index = len(p.responses) - 1
	}
	p.turn++
	return p.responses[index], nil
}

func newTestLoop(t *testing.T, provider types.Provider, tools []types.Tool) (*AgentLoop, *types.ExecutionPolicy) {
	t.Helper()
	dir := t.TempDir()
	session := NewSession(dir, "test")
	if err := session.Init(map[string]any{}); err != nil {
		t.Fatal(err)
	}
	policy := &types.ExecutionPolicy{
		CommandTimeoutMs: 5000, MaxReadBytes: 1024, MaxToolOutputChars: 4000,
		ApprovalMode: types.ApprovalSafe, Approvals: map[string]string{},
	}
	toolContext := &types.ToolContext{Workspace: dir, SessionDir: dir, Policy: policy}
	loop := NewAgentLoop(LoopOptions{
		Config: &types.AgentConfig{
			PlannerProvider: "scripted", ExecutorProvider: "scripted",
			ResearcherProvider: "scripted",
			DefaultRole:        types.RoleAuto, MaxTurns: 6,
			Providers: map[string]*types.ProviderConfig{"scripted": provider.Config()},
		},
		Providers:   map[string]types.Provider{"scripted": provider},
		Tools:       tools,
		ToolContext: toolContext,
		Session:     session,
	})
	return loop, policy
}

func echoTool(calls *int) types.Tool {
	return types.Tool{
		Name: "echo", Description: "echo", Risk: types.RiskRead,
		Parameters: map[string]any{"type": "object"},
		Execute: func(args map[string]any, tc types.ToolContext) (types.ToolResult, error) {
			*calls++
			return types.ToolResult{Content: []types.ContentBlock{types.TextBlock("echoed")}}, nil
		},
	}
}

func TestRunExecutesToolsAndFinishes(t *testing.T) {
	provider := &scriptedProvider{
		config: &types.ProviderConfig{Type: types.KindMock, Model: "scripted"},
		responses: []types.ProviderResponse{
			{ToolCalls: []types.ToolCall{{ID: "1", Name: "echo", Arguments: map[string]any{}}}},
			{Text: "all done", Usage: types.TokenUsage{Output: 12}},
		},
	}
	calls := 0
	loop, _ := newTestLoop(t, provider, []types.Tool{echoTool(&calls)})

	var kinds []string
	result, err := loop.Run("do the thing", RunOptions{
		OnEvent: func(event LoopEvent) { kinds = append(kinds, event.Type) },
	})
	if err != nil {
		t.Fatal(err)
	}
	if calls != 1 {
		t.Fatalf("expected the tool to run once, ran %d times", calls)
	}
	if result.Turns != 2 {
		t.Fatalf("expected 2 turns, got %d", result.Turns)
	}
	if result.Usage.Output != 12 {
		t.Fatalf("usage not accumulated: %+v", result.Usage)
	}
	if !strings.Contains(strings.Join(kinds, ","), "tool_start,tool_end") {
		t.Fatalf("tool events missing: %v", kinds)
	}
	// The tool result has to reach the provider on the next turn.
	if len(provider.seen) < 2 {
		t.Fatal("expected a second request")
	}
	second := provider.seen[1]
	if second[len(second)-1].Role != types.MessageToolResult {
		t.Fatalf("tool result not sent back: %+v", second[len(second)-1])
	}
}

func TestRunReportsMissingTool(t *testing.T) {
	provider := &scriptedProvider{
		config: &types.ProviderConfig{Type: types.KindMock, Model: "scripted"},
		responses: []types.ProviderResponse{
			{ToolCalls: []types.ToolCall{{ID: "1", Name: "nope", Arguments: map[string]any{}}}},
			{Text: "recovered"},
		},
	}
	loop, _ := newTestLoop(t, provider, nil)
	if _, err := loop.Run("go", RunOptions{}); err != nil {
		t.Fatal(err)
	}
	found := false
	for _, message := range loop.History() {
		if message.Role == types.MessageToolResult && strings.Contains(message.Text(), "Tool not found") {
			found = true
		}
	}
	if !found {
		t.Fatal("missing tool did not produce an error result")
	}
}

func TestRunInterruptedKeepsHistoryValid(t *testing.T) {
	provider := &scriptedProvider{
		config: &types.ProviderConfig{Type: types.KindMock, Model: "scripted"},
		responses: []types.ProviderResponse{
			{ToolCalls: []types.ToolCall{
				{ID: "1", Name: "echo", Arguments: map[string]any{}},
				{ID: "2", Name: "echo", Arguments: map[string]any{}},
			}},
		},
	}
	calls := 0
	loop, _ := newTestLoop(t, provider, []types.Tool{echoTool(&calls)})

	ctx, cancel := context.WithCancel(context.Background())
	result, err := loop.Run("go", RunOptions{
		Ctx: ctx,
		OnEvent: func(event LoopEvent) {
			if event.Type == "tool_end" {
				cancel() // interrupt between the two calls
			}
		},
	})
	if err != nil {
		t.Fatal(err)
	}
	if !result.Interrupted {
		t.Fatal("expected the run to report an interrupt")
	}
	// Every issued call must still have a result, or strict chat APIs reject
	// the history on the next turn.
	answered := map[string]bool{}
	for _, message := range result.Messages {
		if message.Role == types.MessageToolResult {
			answered[message.ToolCallID] = true
		}
	}
	for _, message := range result.Messages {
		for _, call := range message.ToolCalls {
			if !answered[call.ID] {
				t.Fatalf("tool call %s dangles after an interrupt", call.ID)
			}
		}
	}
	last := result.Messages[len(result.Messages)-1]
	if last.Role != types.MessageAssistant || !strings.Contains(last.Text(), "interrupted") {
		t.Fatalf("expected an interrupt marker, got %+v", last)
	}
}

func TestRoutingPrefersExecutorForExecutionPrompts(t *testing.T) {
	loop, _ := newTestLoop(t, &scriptedProvider{
		config: &types.ProviderConfig{Type: types.KindMock, Model: "scripted"},
	}, nil)
	loop.options.Config.PlannerProvider = "planner"
	loop.options.Config.ExecutorProvider = "executor"

	if got := loop.routeProvider(types.RoleAuto, "run ls -la in the workspace"); got != "executor" {
		t.Fatalf("expected executor routing, got %s", got)
	}
	if got := loop.routeProvider(types.RoleAuto, "运行 一下这个二进制"); got != "executor" {
		t.Fatalf("expected executor routing for the Chinese keyword, got %s", got)
	}
	if got := loop.routeProvider(types.RoleAuto, "what is the best approach here"); got != "planner" {
		t.Fatalf("expected planner routing, got %s", got)
	}
	if got := loop.routeProvider(types.RoleExecutor, "anything"); got != "executor" {
		t.Fatalf("explicit role ignored: %s", got)
	}
	loop.options.Config.ResearcherProvider = "researcher"
	if got := loop.routeProvider(types.RoleResearcher, "map prior art"); got != "researcher" {
		t.Fatalf("expected researcher routing, got %s", got)
	}
}

func TestResetPlanClearsLiveTaskListOnly(t *testing.T) {
	provider := &scriptedProvider{
		config:    &types.ProviderConfig{Type: types.KindMock, Model: "scripted"},
		responses: []types.ProviderResponse{{Text: "ok"}},
	}
	loop, _ := newTestLoop(t, provider, nil)
	loop.publishPlan([]types.PlanStep{{Text: "old task", Status: types.StepInProgress}},
		types.PlanUpdateMeta{Source: "test"})
	if loop.Plan() == nil {
		t.Fatal("plan was not published")
	}
	loop.ResetPlan()
	if loop.Plan() != nil {
		t.Fatalf("plan should be cleared, got %+v", loop.Plan())
	}
	if _, err := loop.Run("new task", RunOptions{}); err != nil {
		t.Fatal(err)
	}
	if len(loop.History()) == 0 {
		t.Fatal("resetting the plan must not clear transcript history")
	}
}

func TestRunAddsEffectiveRolePrompt(t *testing.T) {
	provider := &scriptedProvider{
		config:    &types.ProviderConfig{Type: types.KindMock, Model: "scripted"},
		responses: []types.ProviderResponse{{Text: "ok"}},
	}
	loop, _ := newTestLoop(t, provider, nil)
	loop.options.SystemPrompt = "global system"
	loop.options.RolePrompts = map[types.AgentRole]string{
		types.RolePlanner:    "planner prompt",
		types.RoleExecutor:   "executor prompt",
		types.RoleResearcher: "researcher prompt",
	}

	if _, err := loop.Run("collect background", RunOptions{Role: types.RoleResearcher}); err != nil {
		t.Fatal(err)
	}
	if len(provider.systems) == 0 || !strings.Contains(provider.systems[0], "researcher prompt") {
		t.Fatalf("researcher prompt missing: %#v", provider.systems)
	}
	if strings.Contains(provider.systems[0], "planner prompt") || strings.Contains(provider.systems[0], "executor prompt") {
		t.Fatalf("wrong role prompt included: %q", provider.systems[0])
	}
}

func TestIsolatedRunUsesOnlyCurrentPromptAndOverrides(t *testing.T) {
	provider := &scriptedProvider{
		config: &types.ProviderConfig{Type: types.KindMock, Model: "scripted"},
		responses: []types.ProviderResponse{
			{Text: "planner saw the full task"},
			{Text: "executor collected local evidence"},
		},
	}
	loop, _ := newTestLoop(t, provider, nil)

	if _, err := loop.Run("full authorized CTF objective for ./chall", RunOptions{Role: types.RolePlanner}); err != nil {
		t.Fatal(err)
	}
	limitedTools := []types.Tool{{Name: "file_info", Description: "identify", Risk: types.RiskRead}}
	if _, err := loop.Run("Objective: collect local evidence about ./chall", RunOptions{
		Role: types.RoleExecutor, Isolated: true, SystemPrompt: "neutral executor system",
		Tools: limitedTools, FreshSession: true,
	}); err != nil {
		t.Fatal(err)
	}

	if len(provider.seen) != 2 {
		t.Fatalf("expected two provider calls, got %d", len(provider.seen))
	}
	second := provider.seen[1]
	if len(second) != 1 {
		t.Fatalf("isolated run should send only the current prompt, got %d messages", len(second))
	}
	if strings.Contains(second[0].Text(), "CTF objective") {
		t.Fatalf("isolated run leaked previous planner prompt: %q", second[0].Text())
	}
	if provider.systems[1] != "neutral executor system" {
		t.Fatalf("system override not applied: %q", provider.systems[1])
	}
	if len(provider.tools[1]) != 1 || provider.tools[1][0].Name != "file_info" {
		t.Fatalf("tool override not applied: %#v", provider.tools[1])
	}
	if !provider.fresh[1] {
		t.Fatal("fresh session flag not forwarded")
	}
}

func TestDescribeEndpoint(t *testing.T) {
	cases := []struct {
		config types.ProviderConfig
		want   string
	}{
		{types.ProviderConfig{Type: types.KindAnthropic}, "https://api.anthropic.com/v1/messages"},
		{types.ProviderConfig{Type: types.KindOpenAIChat, BaseURL: "https://api.deepseek.com/v1/"}, "https://api.deepseek.com/v1/chat/completions"},
		{types.ProviderConfig{Type: types.KindOpenAIResponses}, "https://api.openai.com/v1/responses"},
		{types.ProviderConfig{Type: types.KindCLITmux, CLICommand: "codex"}, "tmux:codex"},
	}
	for _, testCase := range cases {
		if got := DescribeEndpoint(&testCase.config); got != testCase.want {
			t.Fatalf("endpoint for %s: got %s want %s", testCase.config.Type, got, testCase.want)
		}
	}
}

File: src/core/agent-loop.test.ts
import { afterAll, describe, expect, test } from "bun:test";
import * as fs from "node:fs/promises";
import * as os from "node:os";
import * as path from "node:path";
import { AgentLoop } from "./agent-loop";
import { JsonlSession } from "./session";
import { InterruptedError, textBlock } from "../utils";
import type { AgentConfig, AgentTool, ChatProvider, ProviderInput, ProviderResponse, ToolContext } from "../types";

const tmpDirs: string[] = [];

afterAll(async () => {
  await Promise.all(tmpDirs.map(dir => fs.rm(dir, { recursive: true, force: true })));
});

async function scratch(): Promise<string> {
  const dir = await fs.mkdtemp(path.join(os.tmpdir(), "0xaf-loop-"));
  tmpDirs.push(dir);
  return dir;
}

const config: AgentConfig = {
  name: "test",
  plannerProvider: "p",
  executorProvider: "p",
  defaultRole: "auto",
  maxTurns: 4,
  providers: { p: { type: "mock", model: "m" } },
};

function provider(script: Array<(input: ProviderInput) => ProviderResponse | Promise<ProviderResponse>>): ChatProvider {
  let turn = 0;
  return {
    name: "p",
    config: config.providers.p,
    async complete(input) {
      const step = script[Math.min(turn++, script.length - 1)];
      return await step(input);
    },
  };
}

const reply = (text: string): ProviderResponse => ({ text, toolCalls: [] });
const callsTool = (name: string, id = "call_1"): ProviderResponse => ({
  text: "",
  toolCalls: [{ id, name, arguments: {} }],
});

async function makeLoop(
  chat: ChatProvider,
  tools: AgentTool[] = [],
): Promise<{ loop: AgentLoop; context: ToolContext }> {
  const dir = await scratch();
  const session = new JsonlSession(dir, "test");
  await session.init({});
  const context: ToolContext = {
    workspace: dir,
    sessionDir: dir,
    policy: { allowWrites: false, allowNetwork: false, allowSensitive: false, commandTimeoutMs: 1000, maxReadBytes: 1024, maxToolOutputChars: 24_000, approvalMode: "safe", approvals: {} },
  };
  return {
    context,
    loop: new AgentLoop({
      config,
      providers: new Map([["p", chat]]),
      tools,
      toolContext: context,
      systemPrompt: "sys",
      session,
    }),
  };
}

const blockingTool = (started: { resolve?: () => void }): AgentTool => ({
  name: "slow",
  description: "waits for the signal",
  risk: "execute",
  parameters: { type: "object", properties: {} },
  async execute(_args, context) {
    started.resolve?.();
    await new Promise<void>((_ok, fail) => {
      context.signal?.addEventListener("abort", () => fail(new InterruptedError()), { once: true });
    });
    return { content: [textBlock("never")] };
  },
});

describe("AgentLoop context budget", () => {
  test("sends a compacted view while keeping the full history", async () => {
    const seen: number[] = [];
    const chat = provider([
      input => {
        seen.push(input.messages.length);
        return reply("ok");
      },
    ]);
    // A budget this small forces a drop on the second turn.
    chat.config.contextBudgetTokens = 60;
    const { loop } = await makeLoop(chat);
    const events: string[] = [];

    await loop.run("first ".repeat(80));
    await loop.run("second ".repeat(80), { onEvent: event => events.push(event.type) });

    expect(seen[0]).toBe(1);
    // Turn 2 has 3 messages of history but is sent as marker + last exchange.
    expect(loop.history).toHaveLength(4);
    expect(seen[1]).toBeLessThan(3);
    expect(events).toContain("compaction");
  });
});

describe("AgentLoop interruption", () => {
  test("a normal run is unaffected", async () => {
    const { loop } = await makeLoop(provider([() => reply("done")]));
    const result = await loop.run("hi");
    expect(result.interrupted).toBeUndefined();
    expect(result.turns).toBe(1);
    expect(loop.history.at(-1)?.role).toBe("assistant");
  });

  test("an already-aborted signal stops before calling the provider", async () => {
    let calls = 0;
    const { loop } = await makeLoop(provider([() => { calls++; return reply("nope"); }]));
    const result = await loop.run("hi", { signal: AbortSignal.abort() });
    expect(calls).toBe(0);
    expect(result.interrupted).toBe(true);
    expect(result.turns).toBe(0);
  });

  test("a provider abort ends the run and leaves a marker in the history", async () => {
    const controller = new AbortController();
    const { loop } = await makeLoop(
      provider([
        () => {
          controller.abort();
          throw new InterruptedError();
        },
      ]),
    );
    const result = await loop.run("hi", { signal: controller.signal });
    expect(result.interrupted).toBe(true);
    const last = loop.history.at(-1);
    expect(last?.role).toBe("assistant");
    expect(last?.role === "assistant" && last.content[0]?.type === "text" && last.content[0].text).toContain("interrupted");
  });

  test("the signal reaches tools, and every tool call still gets a result", async () => {
    const controller = new AbortController();
    const started: { resolve?: () => void } = {};
    const running = new Promise<void>(resolve => {
      started.resolve = resolve;
    });
    const { loop } = await makeLoop(
      provider([
        () => ({ text: "", toolCalls: [{ id: "a", name: "slow", arguments: {} }, { id: "b", name: "slow", arguments: {} }] }),
        () => reply("unreachable"),
      ]),
      [blockingTool(started)],
    );
    const run = loop.run("hi", { signal: controller.signal });
    await running;
    controller.abort();
    const result = await run;

    expect(result.interrupted).toBe(true);
    const assistant = loop.history.find(message => message.role === "assistant" && message.toolCalls?.length);
    const results = loop.history.filter(message => message.role === "toolResult");
    // Pairing invariant: strict chat APIs reject a dangling tool call.
    expect(assistant?.role === "assistant" && assistant.toolCalls?.length).toBe(2);
    expect(results.map(message => message.role === "toolResult" && message.toolCallId)).toEqual(["a", "b"]);
    expect(results.every(message => message.role === "toolResult" && message.isError)).toBe(true);
  });

  test("interrupting between turns stops the loop", async () => {
    const controller = new AbortController();
    let calls = 0;
    const { loop } = await makeLoop(
      provider([
        () => {
          calls++;
          controller.abort();
          return callsTool("missing");
        },
        () => {
          calls++;
          return reply("should not run");
        },
      ]),
    );
    const result = await loop.run("hi", { signal: controller.signal });
    expect(calls).toBe(1);
    expect(result.interrupted).toBe(true);
    // The unknown tool still produced a paired result before the loop bailed.
    expect(loop.history.filter(message => message.role === "toolResult")).toHaveLength(1);
  });
});
